agents/lakeshore336/LS336_agent.py

```diff
@@ -0,0 +1,782 @@
+            if self.module.channels[channel].get_units() != 'kelvin':
+                session.add_message(f'Setting preferred units to kelvin on heater control input.')
+                self.module.channels[channel].set_units('kelvin')
+
+            # restart setpoint at current temperature
+            current_temp = np.round(float(self.module.get_kelvin(channel)), 4)
+            session.add_message(f'Turning ramp off and setting setpoint to current temperature {current_temp}')
+            heater.set_ramp_on_off('off')
+            heater.set_setpoint(current_temp)
+
+            # reset ramp settings
+            ramp = params.get('ramp', 0.1)
+            session.add_message(f'Turning ramp on and setting rate to {ramp}K/min')
+            heater.set_ramp_on_off('on')
+            heater.set_ramp_rate(ramp)
+
+            # make sure not exceeding channel T limit
+            T_limit = self.module.channels[channel].get_T_limit()
+            if T_limit <= params['temperature']:
+                return False, f"{heater.output_name} control channel {channel} T limit of \
+                     {T_limit}K is higher than setpoint of {params['temperature']}"
+
+            # set setpoint
+            if params['temperature'] == current_setpoint:
+                print('Current setpoint matches commanded value. Proceeding unchanged')
+            else:
+                session.add_message(f"Setting {heater.output_name} setpoint to {params['temperature']}")
+                heater.set_setpoint(params['temperature'])
+                self.static_setpoint = params['temperature'] # static setpoint used in temp stability check to avoid ramping bug
+
+                # if transport, restart control loop when setpoint first crossed
+                if params.get('transport', False):
+                            
+                    current_range = heater.get_heater_range()
+                    starting_sign = np.sign(params['temperature'] - current_temp)
+                    transporting = True
+
+                    # if we are raising temp, allow possibility of stopping transport at a cooler temp
+                    T_offset = 0
+                    if starting_sign > 0:
+                        T_offset = params.get('transport_offset', 0)
+                        if T_offset < 0:
+                            return False, f'Transport offset temperature cannot be negative'
+
+                    while transporting:
+                        current_temp = np.round(float(self.module.get_kelvin(channel)), 4)
+                        current_sign = np.sign(params['temperature'] - T_offset - current_temp) # check when this flips
+                        
+                        # release and reacquire lock between data acquisition
+                        self.lock.release()
+                        time.sleep(t_sample)
+                        if not self.lock.acquire(timeout = 10, job = 'servo_to_temperature'):
```

Comment:
This Task can take a while to run; it's cryogenically limited (eg, around 5min say). As with acq, have replaced this with a return False statement.

Reply:
The code I have should be functionally the same as the release_and_acquire method, except with an added sleep statement between the release and the attempted acquisition. This I think better reflects expected behavior of such a process (same thing for acq): it only attempts to acquire the lock when it needs it (e.g. for reading a temperature, infrequently) and the lock is otherwise "totally free" in the interim for short processes. Since the act of reading a temperature is short, short processes can have short timeouts (e.g. 3 seconds) without having to worry that the release_and_acquire method is called frequently-enough in acq or here (eg, with a dummy variable "last_release" etc).

Reply:
*Short tasks can have short timeouts